collect_app/src/main/java/org/odk/collect/android/formmanagement/FormUpdateChecker.kt

```diff
@@ -0,0 +1,100 @@
+package org.odk.collect.android.formmanagement
+
+import android.content.Context
+import org.javarosa.core.reference.ReferenceManager
+import org.odk.collect.analytics.Analytics
+import org.odk.collect.android.R
+import org.odk.collect.android.backgroundwork.ChangeLock
+import org.odk.collect.android.notifications.Notifier
+import org.odk.collect.android.preferences.keys.GeneralKeys
+import org.odk.collect.android.preferences.source.SettingsProvider
+import org.odk.collect.android.provider.FormsProviderAPI
+import org.odk.collect.android.storage.StoragePathProvider
+import org.odk.collect.android.storage.StorageSubdirectory
+import org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer
+import org.odk.collect.android.utilities.FormsRepositoryProvider
+import org.odk.collect.android.utilities.TranslationHandler
+import org.odk.collect.forms.FormSourceException
+import java.io.File
+import java.util.stream.Collectors
+
+class FormUpdateChecker(
```

Comment:
As discussed on Slack and calls, I'd like to expand on this for other jobs and make it a `FormService`. This would also replace `SyncStatusAppState`.

Reply:
I've looked at the way things are structured in this PR, read [the Slack thread](https://getodk.slack.com/archives/C35ENHA64/p1622453471012200) about this, reviewed [the Miro board](https://miro.com/app/board/o9J_lBsLESw=/) and briefly discussed with @seadowg. I think the big questions are whether the various existing functionality that already exists actually belongs together, whether there's a risk to introducing a class that could end up as a "dumping ground" (see `FormController`) and whether a `Service` could be confusing given the existence of Android `Service`s.

I see the value of replacing disparate classes that do related things with a single class that has all the right dependencies so I'm in favor of moving in that direction. I think it'd be really helpful to try to draw bounds around what this thing will do either in comments, naming or both. Maybe it's `FormUpdateManager` so it's clearly about the various kinds of updates we support. @seadowg you mentioned this new abstraction possibly managing the list of forms on the device. I think a `latestLocalFormList` or whatever makes sense in something like a `FormUpdateManager`.

Reply:
All good points. I'll try and move towards this kind of approach in the next PR (that will look at match exactly) but will stay away from the `XService` because of both the reasons you mention.

Reply:
Sounds great. One thing I want to bring up explicitly so you can have it in your brain as you do that is that I continue to find the use of `sync` in code for the "match exactly" form update mode confusing. It's not a big deal but perhaps this will be an opportunity to make it more clear that "match exactly" is really nothing other than a type of form update. It's not two directional, it doesn't involve submissions, it doesn't involve settings, etc.